### PR TITLE
Powerflex Replication

### PR DESCRIPTION
**Pull request by falfaroc**
# Description
Add the documentation related to enabling `Replication` on PowerFlex.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| https://github.com/dell/csm/issues/618  |

# Checklist:

- [x] Have you run a grammar and spell checks against your submission?
- [x] Have you tested the changes locally?
- [x] Have you tested whether the hyperlinks are working properly?
- [ ] Did you add the examples wherever applicable?
- [ ] Have you added high-resolution images?

### CODE REVIEW

**Comment by rajkumar-palani** [comment]
@falfaroc - Demo link is not working, please check. Can you please also correct the typo "Replication" given in the description.

**Review by rajkumar-palani — COMMENTED** [review]
@falfaroc - Demo link is not working, please check. Can you please also correct the typo (update as "Replication") given in the description.

File: content/docs/replication/_index.md [hunk 1]
```diff
@@ -18,40 +18,40 @@ CSM for Replication provides the following capabilities:
...
 {{<table "table table-striped table-bordered table-sm">}}
-| COP/OS            | PowerMax         | PowerStore       | PowerScale       |
-| ----------------- | ---------------- | ---------------- | ---------------- |
-| Kubernetes        | 1.24, 1.25, 1.26 | 1.22, 1.23, 1.24 | 1.22, 1.23, 1.24 |
-| Red Hat OpenShift | 4.10, 4.11       | 4.9, 4.10        | 4.9, 4.10        |
-| RHEL              | 7.x, 8.x         | 7.x, 8.x         | 7.x, 8.x         |
-| CentOS            | 7.8, 7.9         | 7.8, 7.9         | 7.8, 7.9         |
-| Ubuntu            | 20.04            | 20.04            | 20.04            |
-| SLES              | 15SP4            | 15SP2            | 15SP2            |
+| COP/OS            | PowerMax         | PowerStore       | PowerScale       | PowerFlex        |
```
Comment (rajkumar-palani) [inline content/docs/replication/_index.md:39]: Does replication Supports for 1.25 and 1.26 for Powerflex?

File: content/docs/replication/_index.md [hunk 2]
```diff
@@ -89,11 +90,11 @@ the objects still exist in pairs.
...
 The following matrix provides a list of all supported versions for each Dell Storage product.
 
-| Platforms        | PowerMax                       | PowerStore       | PowerScale       |
-| ---------------- | ------------------------------ | ---------------- | ---------------- |
-| Kubernetes       | 1.24, 1.25, 1.26               | 1.22, 1.23, 1.24 | 1.22, 1.23, 1.24 |
-| RedHat Openshift | 4.10, 4.11                     | 4.9, 4.10        | 4.9, 4.10        |
-| CSI Driver       | 2.x(k8s), <br> 2.2+(OpenShift) | 2.x              | 2.2+             |
+| Platforms        | PowerMax                       | PowerStore       | PowerScale       | PowerFlex        |
+| ---------------- | ------------------------------ | ---------------- | ---------------- | ---------------- |
+| Kubernetes       | 1.24, 1.25, 1.26               | 1.22, 1.23, 1.24 | 1.22, 1.23, 1.24 | 1.24, 1.25       |
```
Comment (shanmydell) [inline content/docs/replication/_index.md:95]: @falfaroc : K8s n and n-2 matrix support has to be qualified

File: content/docs/replication/deployment/installation.md [hunk 3]
```diff
@@ -75,8 +75,10 @@ The following CSI drivers support replication:
 1. CSI driver for PowerMax
 2. CSI driver for PowerStore
 3. CSI driver for PowerScale
+4. CSI driver for Unity XT
```
Comment (shanmydell) [inline content/docs/replication/deployment/installation.md:78]: Unity doesn't support replication

File: content/docs/replication/_index.md [hunk 4]
```diff
@@ -18,40 +18,40 @@ CSM for Replication provides the following capabilities:
...
-| Ubuntu            | 20.04            | 20.04            | 20.04            |
-| SLES              | 15SP4            | 15SP2            | 15SP2            |
+| COP/OS            | PowerMax         | PowerStore       | PowerScale       | PowerFlex        |
+| ----------------- | ---------------- | ---------------- | ---------------- | ---------------- |
+| Kubernetes        | 1.24, 1.25, 1.26 | 1.22, 1.23, 1.24 | 1.22, 1.23, 1.24 | 1.24             |
+| Red Hat OpenShift | 4.10, 4.11       | 4.9, 4.10        | 4.9, 4.10        | 4.9, 4.10        |
+| RHEL              | 7.x, 8.x         | 7.x, 8.x         | 7.x, 8.x         | 8.x              |
+| CentOS            | 7.8, 7.9         | 7.8, 7.9         | 7.8, 7.9         | 7.9              |
+| Ubuntu            | 20.04            | 20.04            | 20.04            | n/a              |
+| SLES              | 15SP4            | 15SP2            | 15SP2            | n/a              |
```
Comment (shanmydell) [inline content/docs/replication/_index.md:46]: Entire matrix needs to be updated

File: content/docs/replication/deployment/powerflex.md [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+## Enabling Replication In CSI PowerFlex
+
+Container Storage Modules (CSM) Replication sidecar is a helper container that
+is installed alongside a CSI driver to facilitate replication functionality.
+Such CSI drivers must implement `dell-csi-extensions` calls.
+
+CSI driver for Dell PowerFlex supports necessary extension calls from
+`dell-csi-extensions`. To be able to provision replicated volumes you would need
+to do the steps described in the following sections.
+
```
Comment (shanmydell) [inline content/docs/replication/deployment/powerflex.md:16]: Are these details with respect to dell-csi-extensions needed in documentation?

File: content/docs/replication/deployment/powerflex.md [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+### Before Installation
+
+#### On Storage Array
+
+Be sure to configure replication between multiple PowerFlex instances using instructions provided by PowerFlex storage.
+
+You can ensure that you configured remote systems by navigating to the `Protection` tab and choosing `Peer Systems` in the UI of your PowerFlex instance.
```
Comment (shanmydell) [inline content/docs/replication/deployment/powerflex.md:23]: Please rephrase sentence and replace 'your'

**Review by shanmydell — CHANGES_REQUESTED** [review]

**Review by rajkumar-palani — DISMISSED** [review]
Complete the Peer Review on the changes done and get the approval.

File: content/docs/replication/deployment/powerflex.md (hunk 5 above)
Comment (falfaroc) [reply content/docs/replication/deployment/powerflex.md:16]: Keeping it consistent with other drivers' replication documentation.

File: content/docs/replication/deployment/powerflex.md [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+Be sure to configure replication between multiple PowerFlex instances using instructions provided by PowerFlex storage.
+
+Ensure that the remote systems are configured by navigating to the `Protection` tab and choosing `Peer Systems` in the UI of the PowerFlex instance.
+
+There should be a list of remote systems with the `State` fields set to `Connected`.
+
+#### In Kubernetes
+Ensure you installed CRDs and replication controller in your clusters.
+
+To verify you have everything in order you can execute the following commands:
```
Comment (rsedlock1958) [inline content/docs/replication/deployment/powerflex.md:30]: Please rewrite this sentence. It does not seem to make sense.

File: content/docs/replication/deployment/powerflex.md [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+  measured in units of time, that may be lost due to a failure.
+* `replication.storage.dell.com/volumeGroupPrefix` represents what string would
+  be appended to the volume group name to differentiate them.
+* `replication.storage.dell.com/consistencyGroupName` represents the desired
+  name to give the consistency group on the PowerFlex array. If omitted, the
+  driver will generate a name for the consistency group.
+* `replication.storage.dell.com/protectionDomain` represents the remote array's
+  protection domain to use.
+
+Let's follow up that with an example, let's assume we have two Kubernetes
```
Comment (rsedlock1958) [inline content/docs/replication/deployment/powerflex.md:135]: Replace comma with a full stop (period "."). Initial uppercase on "Let's" to start a new sentence.

File: content/docs/replication/deployment/powerflex.md [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+clusters and two PowerFlex storage arrays:
+* Clusters have IDs of `cluster-1` and `cluster-2`
+* Cluster `cluster-1` connected to array `000000000001`
+* Cluster `cluster-2` connected to array `000000000002`
+* For `cluster-1` we plan to use storage pool `pool1` and protection domain `domain1`
+* For `cluster-2` we plan to use storage pool `pool1` and protection domain `domain1`
+
+And this how would our pair of storage classes would look like:
```
Comment (rsedlock1958) [inline content/docs/replication/deployment/powerflex.md:143]: Add "is" after "this"

File: content/docs/replication/deployment/powerflex.md [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+To create storage classes with `repctl` you need to fill up the config with
+the necessary information. You can find an example in
+[here](https://github.com/dell/csm-replication/blob/main/repctl/examples/powerflex_example_values.yaml),
+copy it, and modify it to your needs.
+
+If you open this example you can see a lot of similar fields and parameters you
+can modify in the storage class.
+
+Let's use the same example from manual installation and see how the config would
+look like
```
Comment (rsedlock1958) [inline content/docs/replication/deployment/powerflex.md:207]: Delete "like"

**Review by rsedlock1958 — COMMENTED** [review]
I have reviewed this. Several edits are required.

File: content/docs/replication/deployment/installation.md [hunk 11]
```diff
@@ -75,8 +75,9 @@ The following CSI drivers support replication:
 1. CSI driver for PowerMax
 2. CSI driver for PowerStore
 3. CSI driver for PowerScale
+4. CSI driver for PowerFlex
 
-Please follow the steps outlined in [PowerMax](../powermax), [PowerStore](../powerstore) or [PowerScale](../powerscale) pages during the driver installation.
+Please follow the steps outlined in [PowerMax](../powermax), [PowerStore](../powerstore), [PowerScale](../powerscale), or [PowerFlex](../powerflex) pages during the driver installation.
```
Comment (santhoshatdell) [inline content/docs/replication/deployment/installation.md:80]: Similar addition is needed at https://dell.github.io/csm-docs/docs/replication/deployment/storageclasses/#driver-specific-parameters

File: content/docs/replication/deployment/powerflex.md [hunk 12]
```diff
@@ -0,0 +1,255 @@
+---
+title: PowerFlex
+linktitle: PowerFlex
+weight: 6
+description: Enabling Replication feature for CSI PowerFlex
```
Comment (santhoshatdell) [inline content/docs/replication/deployment/powerflex.md:5]: Please update the release notes section about new feature.
https://dell.github.io/csm-docs/docs/replication/release/#new-featureschanges

File: content/docs/replication/deployment/powerflex.md (hunk 8 above)
Comment (santhoshatdell) [reply content/docs/replication/deployment/powerflex.md]: @falfaroc : If possible, could you please correct the grammar mistakes in other driver files as well?

File: content/docs/replication/deployment/powerflex.md [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+  replication.storage.dell.com/isReplicationEnabled: "true"
+  replication.storage.dell.com/remoteStorageClassName: "vxflexos-replication"
+  replication.storage.dell.com/remoteClusterID: <remoteClusterID>
+  replication.storage.dell.com/remoteSystem: <remoteSystemID>
+  replication.storage.dell.com/remoteStoragePool: <remoteStoragePool>
+  replication.storage.dell.com/rpo: 60
+  replication.storage.dell.com/volumeGroupPrefix: "csi"
+  replication.storage.dell.com/consistencyGroupName: <desiredConsistencyGroupName>
+  replication.storage.dell.com/protectionDomain: <remoteProtectionDomain>
+```
```
Comment (santhoshatdell) [inline content/docs/replication/deployment/powerflex.md:112]: arrayID (or systemID), storagePool and protectionDomain are missing.

File: content/docs/replication/deployment/powerflex.md [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+  the array.
+* `replication.storage.dell.com/rpo` is an acceptable amount of data, which is
+  measured in units of time, that may be lost due to a failure.
+* `replication.storage.dell.com/volumeGroupPrefix` represents what string would
+  be appended to the volume group name to differentiate them.
+* `replication.storage.dell.com/consistencyGroupName` represents the desired
+  name to give the consistency group on the PowerFlex array. If omitted, the
+  driver will generate a name for the consistency group.
+* `replication.storage.dell.com/protectionDomain` represents the remote array's
+  protection domain to use.
```
Comment (santhoshatdell) [inline content/docs/replication/deployment/powerflex.md:133]: arrayID (or systemID), storagePool and protectionDomain are missing.

**Review by rsedlock1958 — COMMENTED** [review]
I have reviewed this. It looks good. No grammatical edits are required.

File: content/docs/replication/deployment/powerflex.md [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+parameters:
+  replication.storage.dell.com/isReplicationEnabled: "true"
+  replication.storage.dell.com/remoteStorageClassName: "vxflexos-replication"
+  replication.storage.dell.com/remoteClusterID: "cluster-1"
+  replication.storage.dell.com/remoteSystem: "000000000001"
+  replication.storage.dell.com/remoteStoragePool: pool1
+  replication.storage.dell.com/protectionDomain: domain1
+  replication.storage.dell.com/rpo: 60
+  replication.storage.dell.com/volumeGroupPrefix: "csi"
+  arrayID: "000000000002"
```
Comment (santhoshatdell) [inline content/docs/replication/deployment/powerflex.md:186]: storagePool and protectionDomain are missing.

File: content/docs/replication/deployment/powerflex.md [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+parameters:
+  replication.storage.dell.com/isReplicationEnabled: "true"
+  replication.storage.dell.com/remoteStorageClassName: "vxflexos-replication"
+  replication.storage.dell.com/remoteClusterID: "cluster-2"
+  replication.storage.dell.com/remoteSystem: "000000000002"
+  replication.storage.dell.com/remoteStoragePool: pool1
+  replication.storage.dell.com/protectionDomain: domain1
+  replication.storage.dell.com/rpo: 60
+  replication.storage.dell.com/volumeGroupPrefix: "csi"
+  arrayID: "000000000001"
```
Comment (santhoshatdell) [inline content/docs/replication/deployment/powerflex.md:164]: storagePool and protectionDomain are missing.

File: content/docs/replication/deployment/powerflex.md [hunk 17]
```diff
@@ -0,0 +1,255 @@
...
+[here](https://github.com/dell/csi-powerflex/blob/main/samples/storageclass/vxflexos-replication.yaml).
+
+It will look like this:
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: vxflexos-replication
+provisioner: csi-vxflexos.dellemc.com
+reclaimPolicy: Delete
```
Comment (santhoshatdell) [inline content/docs/replication/deployment/powerflex.md:99]: reclaimPolicy is _Retain_ at other places. Let us be consistent.

File: content/docs/replication/deployment/powerflex.md [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+to installation instructions in [installation-repctl](../install-repctl) or
+[installation](../installation).
+
+### Installing Driver With Replication Module
+
+To install the driver with replication enabled, you need to ensure you have set
+helm parameter `replication.enabled` in your copy of example `values.yaml` file
+(usually called `my-powerflex-settings.yaml`, `myvalues.yaml` etc.).
+
+Here is an example of how that would look like
```
Comment (rsedlock1958) [inline content/docs/replication/deployment/powerflex.md:56]: Delete "like"

**Review by rsedlock1958 — COMMENTED** [review]
I have reviewed this. One edit is required.

**Review by santhoshatdell — APPROVED** [review]

**Comment by falfaroc** [comment]
> I have reviewed this. No edits are required.

@rsedlock1958 Please approve if no further changes are needed.

File: content/docs/replication/_index.md [hunk 19]
```diff
@@ -18,40 +18,40 @@ CSM for Replication provides the following capabilities:
...
-| Kubernetes        | 1.24, 1.25, 1.26 | 1.24, 1.25, 1.26 | 1.24, 1.25, 1.26 |
-| Red Hat OpenShift | 4.10, 4.11       | 4.9, 4.10        | 4.9, 4.10        |
-| RHEL              | 7.x, 8.x         | 7.x, 8.x         | 7.x, 8.x         |
-| CentOS            | 7.8, 7.9         | 7.8, 7.9         | 7.8, 7.9         |
-| Ubuntu            | 20.04            | 20.04            | 20.04            |
-| SLES              | 15SP4            | 15SP2            | 15SP2            |
+| COP/OS            | PowerMax         | PowerStore       | PowerScale       | PowerFlex        |
+| ----------------- | ---------------- | ---------------- | ---------------- | ---------------- |
+| Kubernetes        | 1.24, 1.25, 1.26 | 1.24, 1.25, 1.26 | 1.24, 1.25, 1.26 | 1.24, 1.25, 1.26 |
+| Red Hat OpenShift | 4.10, 4.11       | 4.9, 4.10        | 4.9, 4.10        | 4.9, 4.10        |
```
Comment (rajkumar-palani) [inline content/docs/replication/_index.md:42]: check whether 4.11 and 4.10 support for power flex.

File: content/docs/replication/_index.md [hunk 20]
```diff
@@ -18,40 +18,40 @@ CSM for Replication provides the following capabilities:
...
-|               |                                                 PowerMax                                                 |     PowerStore      |             PowerScale             |
-| ------------- | :------------------------------------------------------------------------------------------------------: | :-----------------: | :--------------------------------: |
-| Storage Array | PowerMax 2000/8000 <br> PowerMax 2500/8500 <br> 5978.479.479, 5978.711.711, 6079.xxx.xxx, Unisphere 10.0 | 1.0.x, 2.0.x, 2.1.x | OneFS 8.1, 8.2, 9.0, 9.1, 9.2, 9.3 |
+|               |                                                 PowerMax                                                 |     PowerStore      |             PowerScale             | PowerFlex |
+| ------------- | :------------------------------------------------------------------------------------------------------: | :-----------------: | :--------------------------------: | :-------: |
+| Storage Array | PowerMax 2000/8000 <br> PowerMax 2500/8500 <br> 5978.479.479, 5978.711.711, 6079.xxx.xxx, Unisphere 10.0 | 1.0.x, 2.0.x, 2.1.x | OneFS 8.1, 8.2, 9.0, 9.1, 9.2, 9.3 | 3.6       |
```
Comment (rajkumar-palani) [inline content/docs/replication/_index.md:54]: do we support replication on other model (3.4, 3.5) of power flex arrays?

File: content/docs/replication/_index.md [hunk 21]
```diff
@@ -89,11 +90,11 @@ the objects still exist in pairs.
...
 
-| Platforms        | PowerMax                       | PowerStore       | PowerScale       |
-| ---------------- | ------------------------------ | ---------------- | ---------------- |
-| Kubernetes       | 1.24, 1.25, 1.26               | 1.24, 1.25, 1.26 | 1.24, 1.25, 1.26 |
-| RedHat Openshift | 4.10, 4.11                     | 4.9, 4.10        | 4.9, 4.10        |
-| CSI Driver       | 2.x(k8s), <br> 2.2+(OpenShift) | 2.x              | 2.2+             |
+| Platforms        | PowerMax                       | PowerStore       | PowerScale       | PowerFlex        |
+| ---------------- | ------------------------------ | ---------------- | ---------------- | ---------------- |
+| Kubernetes       | 1.24, 1.25, 1.26               | 1.24, 1.25, 1.26 | 1.24, 1.25, 1.26 | 1.24, 1.25, 1.26 |
+| RedHat Openshift | 4.10, 4.11                     | 4.9, 4.10        | 4.9, 4.10        | 4.9, 4.10        |
```
Comment (rajkumar-palani) [inline content/docs/replication/_index.md:96]: whether powerflex supports 4.11 and 4.10. please double check?

**Review by rajkumar-palani — DISMISSED** [review]
@harshaatdell - can you please review this PR?

File: content/docs/replication/_index.md (hunk 20 above)
Comment (falfaroc) [reply content/docs/replication/_index.md]: Targetted and verified for PowerFlex 3.6 and 4.0.

**Review by rsedlock1958 — APPROVED** [review]
I have reviewed this. No grammatical edits are required.

**Review by santhoshatdell — APPROVED** [review]

**Review by harshaatdell — APPROVED** [review]

**Comment by harshaatdell** [comment]
> @harshaatdell - can you please review this PR?

Done

**Pull request merged**